File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Storage/PrivateEndpointStorage.swift
//
//  PrivateEndpointStorage.swift
//  PaykitDemo
//
//  Persistent storage for private endpoints using Keychain.
//

import Foundation
// PaykitMobile types are in the same module, no import needed

/// Manages persistent storage of private payment endpoints
class PrivateEndpointStorage {
    
    private let keychain: KeychainStorage
    private let identityName: String
    
    // In-memory cache
    private var endpointsCache: [String: [PrivateEndpointOffer]]?
    
    private var endpointsKey: String {
        "paykit.private_endpoints.\(identityName)"
    }
    
    init(identityName: String, keychain: KeychainStorage = KeychainStorage(serviceIdentifier: "com.paykit.demo")) {
        self.identityName = identityName
        self.keychain = keychain
    }
    
    // MARK: - CRUD Operations
    
    /// Get all private endpoints for a peer
    func listForPeer(_ peerPubkey: String) -> [PrivateEndpointOffer] {
        let all = loadAllEndpoints()
        return all[peerPubkey] ?? []
    }
    
    /// Get a specific endpoint for a peer and method
    func get(peerPubkey: String, methodId: String) -> PrivateEndpointOffer? {
        let endpoints = listForPeer(peerPubkey)
        return endpoints.first { $0.methodId == methodId }
    }
    
    /// Save a private endpoint
    func save(_ endpoint: PrivateEndpointOffer, forPeer peerPubkey: String) throws {
        var all = loadAllEndpoints()
        
        // Get or create list for this peer
        var peerEndpoints = all[peerPubkey] ?? []
        
        // Remove existing endpoint for this method if it exists
        peerEndpoints.removeAll { $0.methodId == endpoint.methodId }
        
        // Add the new endpoint
        peerEndpoints.append(endpoint)
        
        // Update the dictionary
        all[peerPubkey] = peerEndpoints
        
        try persistAllEndpoints(all)
    }
    
    /// Remove a specific endpoint
    func remove(peerPubkey: String, methodId: String) throws {
        var all = loadAllEndpoints()
        
        guard var peerEndpoints = all[peerPubkey] else {
            return // Nothing to remove
        }
        
        peerEndpoints.removeAll { $0.methodId == methodId }
        
        if peerEndpoints.isEmpty {
            all.removeValue(forKey: peerPubkey)
        } else {
            all[peerPubkey] = peerEndpoints
        }
        
        try persistAllEndpoints(all)
    }
    
    /// Remove all endpoints for a peer
    func removeAllForPeer(_ peerPubkey: String) throws {
        var all = loadAllEndpoints()
        all.removeValue(forKey: peerPubkey)
        try persistAllEndpoints(all)
    }
    
    /// List all peers that have private endpoints
    func listPeers() -> [String] {
        let all = loadAllEndpoints()
        return Array(all.keys)
    }
    
    /// Clean up expired endpoints
    /// Note: Expiration checking would need to be added to PrivateEndpointOffer
    /// For now, this is a placeholder
    func cleanupExpired() -> Int {
        // TODO: Implement expiration checking when PrivateEndpointOffer includes expiresAt
        return 0
    }
    
    /// Get count of all stored endpoints
    func count() -> Int {
        let all = loadAllEndpoints()
        return all.values.reduce(0) { $0 + $1.count }
    }
    
    /// Clear all endpoints
    func clearAll() throws {
        try persistAllEndpoints([:])
    }
    
    // MARK: - Private Helpers
    
    private func loadAllEndpoints() -> [String: [PrivateEndpointOffer]] {
        if let cached = endpointsCache {
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: endpointsKey) else {
                return [:]
            }
            
            // Decode from JSON
            let decoder = JSONDecoder()
            let stored = try decoder.decode(StoredEndpoints.self, from: data)
            
            // Convert stored format to PrivateEndpointOffer
            var result: [String: [PrivateEndpointOffer]] = [:]
            for (peer, endpoints) in stored.endpoints {
                result[peer] = endpoints.map { stored in
                    PrivateEndpointOffer(
                        methodId: stored.methodId,
                        endpoint: stored.endpoint
                    )
                }
            }
            
            endpointsCache = result
            return result
        } catch {
            print("PrivateEndpointStorage: Failed to load endpoints: \(error)")
            return [:]
        }
    }
    
    private func persistAllEndpoints(_ endpoints: [String: [PrivateEndpointOffer]]) throws {
        // Convert PrivateEndpointOffer to storable format
        var stored: [String: [StoredEndpoint]] = [:]
        for (peer, offers) in endpoints {
            stored[peer] = offers.map { offer in
                StoredEndpoint(
                    methodId: offer.methodId,
                    endpoint: offer.endpoint
                )
            }
        }
        
        let storedEndpoints = StoredEndpoints(endpoints: stored)
        let encoder = JSONEncoder()
        let data = try encoder.encode(storedEndpoints)
        
        try keychain.store(key: endpointsKey, data: data)
        endpointsCache = endpoints
    }
}

// MARK: - Storage Models

private struct StoredEndpoints: Codable {
    var endpoints: [String: [StoredEndpoint]]
}

private struct StoredEndpoint: Codable {
    let methodId: String
    let endpoint: String
}

// MARK: - Errors

enum PrivateEndpointStorageError: LocalizedError {
    case encodingFailed
    case decodingFailed
    case notFound(peer: String, methodId: String)
    
    var errorDescription: String? {
        switch self {
        case .encodingFailed:
            return "Failed to encode private endpoints"
        case .decodingFailed:
            return "Failed to decode private endpoints"
        case .notFound(let peer, let methodId):
            return "Private endpoint not found for peer \(peer), method \(methodId)"
        }
    }
}

File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/storage/PrivateEndpointStorage.kt
package com.paykit.demo.storage

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.paykit.mobile.PrivateEndpointOffer
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.decodeFromString

/**
 * Manages persistent storage of private payment endpoints using EncryptedSharedPreferences.
 */
class PrivateEndpointStorage(context: Context, private val identityName: String) {
    
    companion object {
        private const val TAG = "PrivateEndpointStorage"
    }
    
    private val PREFS_NAME = "paykit_private_endpoints_$identityName"
    private val ENDPOINTS_KEY = "endpoints_map"
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        encodeDefaults = true
    }
    
    private val prefs by lazy {
        try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            
            EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create encrypted prefs, falling back to regular prefs", e)
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        }
    }
    
    // In-memory cache
    private var endpointsCache: Map<String, List<PrivateEndpointOffer>>? = null
    
    // MARK: - CRUD Operations
    
    /**
     * Get all private endpoints for a peer
     */
    fun listForPeer(peerPubkey: String): List<PrivateEndpointOffer> {
        val all = loadAllEndpoints()
        return all[peerPubkey] ?: emptyList()
    }
    
    /**
     * Get a specific endpoint for a peer and method
     */
    fun get(peerPubkey: String, methodId: String): PrivateEndpointOffer? {
        val endpoints = listForPeer(peerPubkey)
        return endpoints.firstOrNull { it.methodId == methodId }
    }
    
    /**
     * Save a private endpoint
     */
    fun save(endpoint: PrivateEndpointOffer, peerPubkey: String) {
        val all = loadAllEndpoints().toMutableMap()
        
        // Get or create list for this peer
        val peerEndpoints = all[peerPubkey]?.toMutableList() ?: mutableListOf()
        
        // Remove existing endpoint for this method if it exists
        peerEndpoints.removeAll { it.methodId == endpoint.methodId }
        
        // Add the new endpoint
        peerEndpoints.add(endpoint)
        
        // Update the map
        all[peerPubkey] = peerEndpoints
        
        persistAllEndpoints(all)
    }
    
    /**
     * Remove a specific endpoint
     */
    fun remove(peerPubkey: String, methodId: String) {
        val all = loadAllEndpoints().toMutableMap()
        
        val peerEndpoints = all[peerPubkey]?.toMutableList() ?: return
        
        peerEndpoints.removeAll { it.methodId == methodId }
        
        if (peerEndpoints.isEmpty) {
            all.remove(peerPubkey)
        } else {
            all[peerPubkey] = peerEndpoints
        }
        
        persistAllEndpoints(all)
    }
    
    /**
     * Remove all endpoints for a peer
     */
    fun removeAllForPeer(peerPubkey: String) {
        val all = loadAllEndpoints().toMutableMap()
        all.remove(peerPubkey)
        persistAllEndpoints(all)
    }
    
    /**
     * List all peers that have private endpoints
     */
    fun listPeers(): List<String> {
        val all = loadAllEndpoints()
        return all.keys.toList()
    }
    
    /**
     * Clean up expired endpoints
     * Note: Expiration checking would need to be added to PrivateEndpointOffer
     * For now, this is a placeholder
     */
    fun cleanupExpired(): Int {
        // TODO: Implement expiration checking when PrivateEndpointOffer includes expiresAt
        return 0
    }
    
    /**
     * Get count of all stored endpoints
     */
    fun count(): Int {
        val all = loadAllEndpoints()
        return all.values.sumOf { it.size }
    }
    
    /**
     * Clear all endpoints
     */
    fun clearAll() {
        persistAllEndpoints(emptyMap())
    }
    
    // MARK: - Private Helpers
    
    private fun loadAllEndpoints(): Map<String, List<PrivateEndpointOffer>> {
        endpointsCache?.let { return it }
        
        return try {
            val jsonString = prefs.getString(ENDPOINTS_KEY, null) ?: return emptyMap()
            val stored = json.decodeFromString<StoredEndpoints>(jsonString)
            
            // Convert stored format to PrivateEndpointOffer
            val result = stored.endpoints.mapValues { (_, storedList) ->
                storedList.map { stored ->
                    PrivateEndpointOffer(
                        methodId = stored.methodId,
                        endpoint = stored.endpoint
                    )
                }
            }
            
            endpointsCache = result
            result
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load endpoints: ${e.message}")
            emptyMap()
        }
    }
    
    private fun persistAllEndpoints(endpoints: Map<String, List<PrivateEndpointOffer>>) {
        try {
            // Convert PrivateEndpointOffer to storable format
            val stored = endpoints.mapValues { (_, offers) ->
                offers.map { offer ->
                    StoredEndpoint(
                        methodId = offer.methodId,
                        endpoint = offer.endpoint
                    )
                }
            }
            
            val storedEndpoints = StoredEndpoints(endpoints = stored)
            val jsonString = json.encodeToString(storedEndpoints)
            
            prefs.edit().putString(ENDPOINTS_KEY, jsonString).apply()
            endpointsCache = endpoints
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save endpoints: ${e.message}")
        }
    }
}

// MARK: - Storage Models

@Serializable
private data class StoredEndpoints(
    val endpoints: Map<String, List<StoredEndpoint>>
)

@Serializable
private data class StoredEndpoint(
    val methodId: String,
    val endpoint: String
)

